Adding DynamoDB high availability features

mcowgill-stripe opened the pull request:
### Context

The main purpose of this change is to enable the ability for external consumers to leverage high availability features without needing ZooKeeper and to be able to do that with configuration instead of needing to modify source code or re-implement logic that is available here today.

The first part of this is creating an extension to use DynamoDB for leader election and leader monitoring. We had submitted prior work from @crioux-stripe and myself to add a KV Store for DynamoDB. That started to divide the extension for a "store" project. Looking at what we'd have to add to separate the store from the leader election/monitoring it makes more sense to create an extension for just DynamoDB and include the store, leader election, and leader monitoring. In theory the only difference is the size of the dependencies and any of the three bring with it roughly the same size. However, if the maintainers have different opinions I'm happy to pivot.

Looking at the code today and projects/docs. The idea seems to be that mantis-control-plane-core is a common dependency between the server and client. This does have exceptions, but seems like it's the intent. With that in mind I designed to have the DynamoDB extension only takea dependency on core and I deprecated and moved things into core to enable that. If that's not the goal/intent of the design today, I'm happy to move all that back. 

In order to finish this goal of providing a way to inject dependencies for DynamoDB there are additional changes that need to happen in MasterMain.java for mantis server and `HighAvailabilityServices` inside of the client configurations. The latter one is more blocking for our internal use case today, since it's harder to unwind the different client paths than it is to re-implement MasterMain. 

In order to make sure I don't go down paths that the maintainers won't accept, I'm opening the PR with the initial DynamoDB code for leader election and leader monitoring. The minimal deprecations and refactoring needed to manage the dependency flow. I am only commenting near the code blocks to add injection. Should I continue in this PR for those sections or should I try to finish and merge this and return to refactor the design for those two code paths (MasterMain and HighAvailabilityServicesUtil).

### Checklist

- [x] `./gradlew build` compiles code correctly
- [x] Added new tests where applicable
- [x] `./gradlew test` passes all tests
- [x] Extended README or added javadocs where applicable

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-store/mantis-control-plane-store-dynamodb/src/test/java/io/mantisrx/server/master/store/DynamoStoreTest.java:
```diff
@@ -77,14 +78,14 @@ public static void setupDynamo() throws Exception {
 			.endpointOverride(URI.create("http://localhost:" + port))
 			.build();
 
+            server.start();
```
I can revert this change.

Andyz26 commented:
@mcowgill-stripe overll lgtm. I see you still have some todo/comments so ping me when you think this is ready for merge.

mcowgill-stripe commented:
I am not sure if I ran into a class erasure issue with the naming or not. After finding that this [commit](https://github.com/Netflix/mantis/pull/675/commits/c3ca083ca8d6aec7fd08639bdc774b245b021c69) passed locally, I chose to rename the KeyValueStore to IKeyValueStore to reduce that possibility.

Andyz26 commented:
> @Andyz26 - I've broken something in container tests it appears. It's not obvious to me what it is yet. I am trying to slowly walk things backwards until I can make this test pass again. If you have a suggestion that would probably accelerate me. If not I'll keep trying to bisect the change.
> 
> Regarding the TODO's this PR had reached a size where I wanted to pause and check in. We could merge these changes and open a new PR to complete those TODOs (assuming I fix the test issue above). Or I could finish proposing the change to enable injection for the high availability services.
> 
> Do you have a preference on the continuing this PR or trying to reach a merge point and separating into a new PR?

IMO, as long as nothing is broken, it's up to you whether you prefer one PR or multiple.
For test errors, the test container can be debugged if you put a breakpoint or something and examine the error in your local Docker app, too.

Andyz26 commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/CoreConfiguration.java:
```diff
@@ -78,4 +77,12 @@ public interface CoreConfiguration {
     int getAsyncHttpClientReadTimeoutMs();
+
+    @Config("mantis.leader.elector.factory")
+    @Default("io.mantisrx.server.core.master.LocalLeaderFactory")
```
A few issues I see here:
* This leaderElector is only for control plane so it doesn't need to be in coreConfig (and all the workerConfig related classes) for worker/agents.
* We are actually trying to deprecate/move away from config-injected objects. In this case the if the electorFactory is only used in control-plane startup I assume it can be either a string in masterConfig directly.

mcowgill-stripe replied:
Happy to remove this part, I noticed it today as well.

I'm not a huge fan of the DI with Skife today (and it's a little out of date). So I'm happy to hear that position.

What are the current thoughts on how we should think about extending Mantis then? I started with the view point I should only take a dependency on core, this way client or server could later bundle my jar. The only option I could find was the object-injection with Skife today. Should I be looking for another pattern here? 

Potentially we could separate each component (@crioux-stripe and I started there with store). That felt pretty heavy weight for each store and we duplicated a lot of DynamoDB config, so this time I went with just an extension for DynamoDB in a gradle project. We could separate each feature into subprojects in the Gradle project and choose different dependencies for each subproject:

```
mantis-ext-dynamodb
  * common (common dynamodb config each subject depends on common)
  * server [subproject only depends on mantis-control-plane-server] (contains store and leader elector) 
  * leader-monitor [subproject depends on mantis-control-plane-core] (contains only monitor)
```

The above feels pretty heavy weight, but perhaps more precise? We'd have to document this somewhere as well and it's sort of straightforward, but definitely not clear when you look at the current source code.

Andyz26 replied:
> I started with the view point I should only take a dependency on core, this way client or server could later bundle my jar. The only option I could find was the object-injection with Skife today. Should I be looking for another pattern here?
Are you saying you don't take a dependency on mantis-control-plane-server, so you don't have access to the MasterConfiguration?

For skife injection, we don't have a full plan yet on how to replace it but we are trying to remove existing usage and avoid adding new ones. Reflection with config string is an ok workaround for now IMO, or have some builder/ctor pass the instance from your wrapper module.
I am open to a zoom call etc. too if you want to further discuss this.

mcowgill-stripe replied:
>Are you saying you don't take a dependency on mantis-control-plane-server, so you don't have access to the MasterConfiguration?

There is no issue with that access that I see because later, that just works out with the DI today.

Andyz26 commented on mantis-runtime-loader/src/main/java/io/mantisrx/runtime/loader/config/WorkerConfigurationWritable.java:
```diff
@@ -80,6 +81,9 @@ public class WorkerConfigurationWritable implements WorkerConfiguration {
     MetricsCollector metricsCollector;
 
+    @JsonIgnore
+    ILeaderMonitorFactory leaderMonitorFactory;
```
do you still need these in worker config?

mcowgill-stripe replied:
I think this does have to be here. The interface requires I add the field and I need some method to coerce to the `ILeaderMonitorFactory` interface, the test error is:

```
io.mantisrx.shaded.com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `io.mantisrx.server.core.ILeaderMonitorFactory` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
```

I think just keeping this ignore is within the current pattern of the workaround and least amount of work?

Andyz26 replied:
The tricky part here is that the workconfig re-created from this class won't have the instances injected (as they are not created in the original frmProperties path).
So I think we got two options here:
1. Avoid instances and always rely on string to re-create instance via reflection.
2. manually set the instance for leaderMonitorFactory again at RuntimeTaskImpl (similar to the setMetricsCollector one.

crioux-stripe commented:
Kicking the CI workflow here to see where we're at.

We debugged the failing container test last week. It was an unfortunate issue of our linter bullying Mike into making a variable final that shouldn't have been. Meeting with Kevin here at Stripe shortly to hand-off merging this. It should be a good (though large) onboarding task for Mantis infrastructure.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/store/NoopStore.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.List;
+import java.util.Map;
+
+public class NoopStore implements IKeyValueStore {
```
Extracted without modification from the nested class.

mcowgill-stripe commented on mantis-testcontainers/src/test/java/TestContainerHelloWorld.java:
```diff
@@ -158,6 +158,7 @@ public static void prepareControlPlane() throws Exception {
                 .withNetwork(network)
                 .withNetworkAliases(CONTROL_PLANE_ALIAS)
                 .withExposedPorts(CONTROL_PLANE_API_PORT, DEBUGGER_PORT)
+                .withLogConsumer(out -> log.info("[Control Plane] {}", out.getUtf8String()))
```
happy to revert this, but I did repeatedly need these logs for debugging the failing test.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-client/src/main/java/io/mantisrx/server/master/client/HighAvailabilityServicesUtil.java:
```diff
@@ -121,99 +124,103 @@ protected void startUp() throws Exception {
+
+        public HighAvailabilityServicesImpl(CoreConfiguration configuration) {
+            this.configuration = configuration;
+            if(configuration.getLeaderMonitorFactory() instanceof LocalLeaderFactory) {
```
This is a backwards compatibility check, to enable the injection solution to work with current configuration files. In a major version I would think we should raise in this situation.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-client/src/main/java/io/mantisrx/server/master/client/HighAvailabilityServicesUtil.java:
```diff
@@ -121,99 +124,103 @@ protected void startUp() throws Exception {
+        }
 
-        ResourceClusterGateway currentResourceClusterGateway =
-            new ResourceClusterGatewayClient(clusterID, masterMonitor.getLatestMaster(), configuration);
```
This part of the file is a confusing unified diff, but this is moved not deleted. If you find changes here please flag them as the intent was no logic code change, just moving.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-client/src/main/java/io/mantisrx/server/master/client/TestGetMasterMonitor.java:
```diff
@@ -56,8 +56,7 @@ public static void main(String[] args) {
         final CountDownLatch latch = new CountDownLatch(5);
         StaticPropertiesConfigurationFactory configurationFactory = new StaticPropertiesConfigurationFactory(properties);
         CoreConfiguration config = configurationFactory.getConfig();
-        final CuratorService curatorService = new CuratorService(config);
```
This is a change that might be controversial. The CuratorService and really the CuratorFramework from IMO is really Zookeeper. In order to make the abstraction around Zookeeper I chose to reduce the exposure and knowledge of the curator to inside the ZK leadership functions.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-client/src/main/java/io/mantisrx/server/master/client/config/StaticPropertiesConfigurationFactory.java:
```diff
@@ -27,6 +29,8 @@ public class StaticPropertiesConfigurationFactory implements ConfigurationFactor
         delegate.addCoercible(new MetricsCoercer(props));
+        delegate.addCoercible(new LeaderElectorFactoryCoercer());
+        delegate.addCoercible(new LeaderMonitorFactoryCoercer());
 //        delegate.addCoercible(new MantisPropertiesCoercer(props));
```
should this be deleted? As I review the PR it's pops out at me, but during my work I didn't notice this commented code.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-client/src/main/java/io/mantisrx/server/master/client/config/StaticPropertiesConfigurationFactory.java:
```diff
@@ -27,6 +29,8 @@ public class StaticPropertiesConfigurationFactory implements ConfigurationFactor
     public StaticPropertiesConfigurationFactory(Properties props) {
         delegate = new ConfigurationObjectFactory(props);
         delegate.addCoercible(new MetricsCoercer(props));
+        delegate.addCoercible(new LeaderElectorFactoryCoercer());
```
tried to avoid these, but our skife config-magic version requires I create these.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/IKeyValueStore.java:
```diff
@@ -0,0 +1,224 @@
+ * mantis-control-plane-server project are example implementations for file,
+ * in-memory and no-op implementations.
+ */
+public interface IKeyValueStore {
```
Move and rename only.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/ILeaderMonitorFactory.java:
```diff
@@ -0,0 +1,23 @@
+
+import io.mantisrx.server.core.master.MasterMonitor;
+
+public interface ILeaderMonitorFactory {
```
I separated the interface for monitor and leader elector to match the split usage.  Logically this did make sense to me, but I'm open to adjusting this.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/ILeadershipManager.java:
```diff
@@ -0,0 +1,54 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.mantisrx.server.core;
```
move and rename only

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/LocalMasterMonitor.java:
```diff
@@ -41,4 +41,19 @@ public Observable<MasterDescription> getMasterObservable() {
     }
+
+    @Override
+    public void start() {
```
This is another potential controversial change. The master monitor was a service like class because the curator was started outside of it, and the curator was a service. When I made the choice to reduce the exposure of the curator to zookeeper the need to start/start (service manage) the monitor needed to move as well. It made sense to me for this to be a service.

However, I can't make it a `BaseService` without making an abstract class. I stopped at implementing the service interface, but not a strongly held position.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java:
```diff
@@ -36,36 +37,45 @@
  * A monitor that monitors the status of Mantis masters.
  */
-public class ZookeeperMasterMonitor extends AbstractIdleService implements MasterMonitor {
+public class ZookeeperMasterMonitor extends BaseService implements MasterMonitor {
```
Moving to `BaseService` makes this feel more consistent, but if the maintainers were moving away from `BaseService` to `AbstractIdleService` then I need to reverse my choices here. In addition, this is where I felt the urge to change the monitor factory interface and ultimately the `MasterMonitor` interface to become an abstract class of `BaseService`. At the moment I'm satisfied with the choice to stop at an interface versus an abstract class.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java:
```diff
@@ -110,19 +120,27 @@ public Observable<MasterDescription> getMasterObservable() {
     @Nullable
     public MasterDescription getLatestMaster() {
-        Preconditions.checkState(isRunning(), "ZookeeperMasterMonitor is currently not running but instead is at state %s", state());
+        Preconditions.checkState(curator.getCurator().getState() == CuratorFrameworkState.STARTED, "ZookeeperMasterMonitor is currently not running but instead is at state %s", curator.getCurator().getState());
```
Double check this, but I believe this is the equivalent logic.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/utils/ConfigUtils.java:
```diff
@@ -64,6 +74,35 @@ public static Method getMethodFromName(String name, Class<? extends CoreConfigur
+     * @return Properties with overrides from the environment variables
+     * @throws RuntimeException if the file can't be found
+     */
+    public static Properties loadProperties(String propFile) {
```
I moved this out of `MasterMain` so I could reuse it in the dynamodb extension work.

When config is refactor, we may want to document precedence (props file, java system, system env)?

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:
```diff
@@ -64,10 +61,6 @@ public CuratorService(CoreConfiguration configs) {
                 .connectString(configs.getZkConnectionString())
                 .build();
-
-        masterMonitor = new ZookeeperMasterMonitor(
```
For me this was a convenience method that was no longer convenient. In my mental model the ZK master monitor needs a curator framework, but the curator doesn't necessarily need to know about the master monitor.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:
```diff
@@ -90,10 +83,11 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
-            setupCuratorListener();
-            curator.start();
-            masterMonitor.startAsync().awaitRunning();
+            if(curator.getState() != CuratorFrameworkState.STARTED) {
```
Starting the curator more than once throws an error. The same will happen if it's in `STOPPED`. I chose to check for not started because if you are stopped and want to be in the running state, that is not possible and I felt an exception should be thrown for that.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBConfig.java:
```diff
@@ -0,0 +1,70 @@
+ */
+package io.mantisrx.extensions.dynamodb;
+
+import org.skife.config.Config;
```
I'm only using this config framework for consistency with upstream. Open to any future changes.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java:
```diff
@@ -0,0 +1,151 @@
+
+    @Override
+    public void shutdown() {
+        logger.info("shutting down");
```
I debated this shutdown order a few times. This is difficult to test, but I assume since leaders kill themselves the procedure can be best effort.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java:
```diff
@@ -0,0 +1,151 @@
+        } catch (RuntimeException | InterruptedException | JsonProcessingException e) {
+            logger.error("leader elector task has failed it will restart, if this error is frequent there is likely a problem with DynamoDB based leader election", e);
+        } finally {
+            isLeaderElectorRunning.set(false);
```
This block is really about restarting the thread. The function to become leader should keep running until it can become the leader. However, exceptions can occur and if they do we should make a best effort to restart. This is a point where I'm unsure about how to handle metrics in new features. Bringing in the logic for the metrics from upstream felt awkward. In the end I used logging for observability. Metric extension is probably a worthy topic for discussion.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java:
```diff
@@ -0,0 +1,151 @@
+        } finally {
+            isLeaderElectorRunning.set(false);
+            if (shouldLeaderElectorBeRunning.get()) {
+                this.leaderElector.schedule(this::tryToBecomeLeader, 1L, TimeUnit.SECONDS);
```
hard coded 1s to restart, do we need a configurable here?

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBMasterMonitor.java:
```diff
@@ -0,0 +1,178 @@
+import rx.subjects.BehaviorSubject;
+
+@Slf4j
+public class DynamoDBMasterMonitor extends BaseService implements MasterMonitor {
```
This design is a simple polling design to monitor the key in DynamoDB. After I finished this code, I know it can be vastly improved. I chose to not push further because there is potential to extract this. Leader monitoring could be completely separate from leader election. In the current implementation ZK and DynamoDB leader monitoring is very knowledgable about how leader is elected. If we change the design to the leader publishes when elected, and heartbeats regularly on a common messaging framework, then monitoring can be separated. This would push requirements to leader election, but I think we could actually push those to leadership manager. That refactor I thought should be a completely separate PR and this implementation should meet the current need.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBStore.java:
```diff
@@ -0,0 +1,282 @@
+import software.amazon.awssdk.services.dynamodb.model.*;
+
+@Slf4j
+public class DynamoDBStore implements IKeyValueStore {
```
rename and not modified.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/resources/dynamodb.properties:
```diff
@@ -0,0 +1,8 @@
+# mantis.ext.dynamodb.store.table=mantis-dynamodb-kv
```
This is an example file, but since the config framework has defaults I left them commented out.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/test/java/io/mantisrx/extensions/dynamodb/DynamoDBClientSingletonTest.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    /**
+     * The use of global state here and the fact the helper rules all start and stop servers around
```
Do we want to yank this out or is my trade off a good choice?

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/test/java/io/mantisrx/extensions/dynamodb/DynamoDBStoreTest.java:
```diff
@@ -0,0 +1,172 @@
+    public static final String V1 = "value1";
+    public static final String V4 = "value4";
+
+    @ClassRule public static DynamoDBLocalRule dynamoDb = new DynamoDBLocalRule();
```
Moved with only minor modifications to fit the Rule patterns.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/zk/ZookeeperLeaderElector.java:
```diff
@@ -0,0 +1,141 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ZookeeperLeaderElector extends BaseService {
```
The original purpose was name change only, but that did grow as the `CuratorService` point of view required modifications to the constructor. I removed the builder model as I felt the factory eliminated the need for it. I did not see wide usage of that pattern and I left `LeaderElector` if users still want it.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/zk/ZookeeperLeadershipFactory.java:
```diff
@@ -0,0 +1,46 @@
+import io.mantisrx.server.master.config.MasterConfiguration;
+import io.mantisrx.shaded.org.apache.curator.utils.ZKPaths;
+
+public class ZookeeperLeadershipFactory extends ZookeeperLeaderMonitorFactory implements ILeaderElectorFactory {
```
This felt awkward to write, but was required since the interface is split and core needed a monitor factory.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:
```diff
@@ -103,17 +96,14 @@ public class MasterMain implements Service {
     private final AtomicBoolean shutdownInitiated = new AtomicBoolean(false);
     private KeyValueBasedPersistenceProvider storageProvider;
-    private CountDownLatch blockUntilShutdown = new CountDownLatch(1);
-    private volatile CuratorService curatorService = null;
```
This becomes a singleton in the factory. I believe that is safe, but worth noting.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:
```diff
@@ -103,17 +96,14 @@ public class MasterMain implements Service {
+    private final CountDownLatch blockUntilShutdown = new CountDownLatch(1);
     private MasterConfiguration config;
     private ILeadershipManager leadershipManager;
-    private final MantisPropertiesLoader dynamicPropertiesLoader;
```
there are cleanups where I removed things that appear dead. I am happy to start reverting these.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:
```diff
@@ -204,17 +194,31 @@ public MasterMain(
+            // set up leader election
+            final ILeaderElectorFactory leaderFactory;
+            final MasterMonitor monitor;
+            if(!config.isLocalMode() && config.getLeaderElectorFactory() instanceof LocalLeaderFactory) {
```
Critical injection point is here. The if/else is only necessary because I wanted to support backwards compatibility with configs that exist today. I thought if I broke configs as they are today we would need a major version bump. I tested with the old config and then modified the properties files in the container test to make sure it worked. I left the source code updated for the new property configuration.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:
```diff
@@ -224,53 +228,6 @@ public MasterMain(
         }
     }
 
-    private static Properties loadProperties(String propFile) {
```
moved to config utils for common reuse.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:
```diff
@@ -346,20 +286,6 @@ public MasterConfiguration getConfig() {
         return config;
     }
 
-    public String getDescriptionJson() {
```
are these safe to delete? I didn't find issues in testing.

mcowgill-stripe commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:
```diff
@@ -329,13 +276,6 @@ public void shutdown() {
             logger.info("Shutting down Mantis Master");
             mantisServices.shutdown();
             logger.info("mantis services shutdown complete");
-            boolean shutdownCuratorEnabled = ConfigurationProvider.getConfig().getShutdownCuratorServiceEnabled();
```
this is moved into the Zookeeper Leader Elector.

Andyz26 commented on mantis-control-plane/mantis-control-plane-client/src/main/java/io/mantisrx/server/master/client/HighAvailabilityServicesUtil.java:
```diff
@@ -121,99 +123,103 @@ protected void startUp() throws Exception {
+
+        public HighAvailabilityServicesImpl(CoreConfiguration configuration) {
+            this.configuration = configuration;
+            if(configuration.getLeaderMonitorFactory() instanceof LocalLeaderFactory) {
+                log.warn("using default non-local Zookeeper leader monitoring you should set: "+
+                    "mantis.leader.monitor.factory=io.mantisrx.server.core.master.ZookeeperLeaderMonitorFactory");
```
I am not sure I follow the logic here: if it's a local leader factory, then use ZookeeperLeaderMonitorFactory? Shouldn't local leader refer to a single node local setup?

mcowgill-stripe commented:
@Andyz26 - I've removed the class based injection and reverted to a simple string. Locally, I have the build passing. Are there other places I need to make changes?

Andyz26 commented:
> @Andyz26 - I've removed the class based injection and reverted to a simple string. Locally, I have the build passing. Are there other places I need to make changes?

hi @mcowgill-stripe did you get a chance to look at the discussion [thread](https://github.com/Netflix/mantis/discussions/677#discussioncomment-9732402) from @sundargates. He made some proposals regarding these interfaces.

crioux-stripe commented:
@Andyz26 @sundargates How would we feel about merging this, and we can have someone take an immediate task to do the interface refactor? This is the last blocker for us going to production before the end of Q2. We can work on the prod rollout and the refactor at the same time.

Andyz26 commented:
> @Andyz26 @sundargates How would we feel about merging this, and we can have someone take an immediate task to do the interface refactor? This is the last blocker for us going to production before the end of Q2. We can work on the prod rollout and the refactor at the same time.

sure that's fine. Let me try a snapshot build to do some validation internally.

Andyz26 commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/CoreConfiguration.java:
```diff
@@ -77,4 +77,8 @@ public interface CoreConfiguration {
+
+    @Config("mantis.leader.monitor.factory")
+    @Default("io.mantisrx.server.core.master.LocalLeaderFactory")
+    String getLeaderMonitorFactory();
```
nit: getLeaderMonitorFactoryName

crioux-stripe commented:
The snapshot error is interesting:

https://github.com/Netflix/mantis/blob/master/build.gradle#L163

```
[Fatal Error] :1:1: Content is not allowed in prolog.
[Fatal Error] :1:1: Content is not allowed in prolog.
[Fatal Error] :1:1: Content is not allowed in prolog.
[Fatal Error] :1:1: Content is not allowed in prolog.
[Fatal Error] :1:1: Content is not allowed in prolog.
[Fatal Error] :1:1: Content is not allowed in prolog.
[Fatal Error] :1:1: Content is not allowed in prolog.
[Fatal Error] :1:1: Content is not allowed in prolog.

FAILURE: Build completed with 36 failures.

1: Task failed with an exception.
-----------
* Where:
Build file '/home/runner/work/mantis/mantis/build.gradle' line: 163

* What went wrong:
Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.

* Try:
> Run with --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
```

Looking into this.

Andyz26 commented:
> The snapshot error is interesting:
> 
> https://github.com/Netflix/mantis/blob/master/build.gradle#L163
> 
> ```
> [Fatal Error] :1:1: Content is not allowed in prolog.
> [Fatal Error] :1:1: Content is not allowed in prolog.
> [Fatal Error] :1:1: Content is not allowed in prolog.
> [Fatal Error] :1:1: Content is not allowed in prolog.
> [Fatal Error] :1:1: Content is not allowed in prolog.
> [Fatal Error] :1:1: Content is not allowed in prolog.
> [Fatal Error] :1:1: Content is not allowed in prolog.
> [Fatal Error] :1:1: Content is not allowed in prolog.
> 
> FAILURE: Build completed with 36 failures.
> 
> 1: Task failed with an exception.
> -----------
> * Where:
> Build file '/home/runner/work/mantis/mantis/build.gradle' line: 163
> 
> * What went wrong:
> Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> > org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
> 
> * Try:
> > Run with --debug option to get more log output.
> 
> * Exception is:
> org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> ```
> 
> Looking into this.

this seems to be only happening on forked repo's branch. I've added both of you to the collaborator list so you can try push ing the branch in the main repo and snapshot build from there.

kmg-stripe commented:
> > The snapshot error is interesting:
> > https://github.com/Netflix/mantis/blob/master/build.gradle#L163
> > ```
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > [Fatal Error] :1:1: Content is not allowed in prolog.
> > 
> > FAILURE: Build completed with 36 failures.
> > 
> > 1: Task failed with an exception.
> > -----------
> > * Where:
> > Build file '/home/runner/work/mantis/mantis/build.gradle' line: 163
> > 
> > * What went wrong:
> > Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> > > org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
> > 
> > * Try:
> > > Run with --debug option to get more log output.
> > 
> > * Exception is:
> > org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looking into this.
> 
> this seems to be only happening on forked repo's branch. I've added both of you to the collaborator list so you can try push ing the branch in the main repo and snapshot build from there.

@Andyz26 - @mcowgill-stripe and @crioux-stripe are out for the next few weeks.  to ensure we make progress on this, i am stepping in to help out here.  could you please add me to the collaborator list?  thanks!

Andyz26 commented:
> > > The snapshot error is interesting:
> > > https://github.com/Netflix/mantis/blob/master/build.gradle#L163
> > > ```
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > 
> > > FAILURE: Build completed with 36 failures.
> > > 
> > > 1: Task failed with an exception.
> > > -----------
> > > * Where:
> > > Build file '/home/runner/work/mantis/mantis/build.gradle' line: 163
> > > 
> > > * What went wrong:
> > > Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> > > > org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
> > > 
> > > * Try:
> > > > Run with --debug option to get more log output.
> > > 
> > > * Exception is:
> > > org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Looking into this.
> > 
> > 
> > this seems to be only happening on forked repo's branch. I've added both of you to the collaborator list so you can try push ing the branch in the main repo and snapshot build from there.
> 
> @Andyz26 - @mcowgill-stripe and @crioux-stripe are out for the next few weeks. to ensure we make progress on this, i am stepping in to help out here. could you please add me to the collaborator list? thanks!

@kmg-stripe could you ask cody to send me an email to verify your github id? Sorry just to be cautious as the write permission is pretty powerful.

kmg-stripe commented:
> > > > The snapshot error is interesting:
> > > > https://github.com/Netflix/mantis/blob/master/build.gradle#L163
> > > > ```
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > [Fatal Error] :1:1: Content is not allowed in prolog.
> > > > 
> > > > FAILURE: Build completed with 36 failures.
> > > > 
> > > > 1: Task failed with an exception.
> > > > -----------
> > > > * Where:
> > > > Build file '/home/runner/work/mantis/mantis/build.gradle' line: 163
> > > > 
> > > > * What went wrong:
> > > > Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> > > > > org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
> > > > 
> > > > * Try:
> > > > > Run with --debug option to get more log output.
> > > > 
> > > > * Exception is:
> > > > org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':mantis-discovery-proto:printReleasedArtifact'.
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Looking into this.
> > > 
> > > 
> > > this seems to be only happening on forked repo's branch. I've added both of you to the collaborator list so you can try push ing the branch in the main repo and snapshot build from there.
> > 
> > 
> > @Andyz26 - @mcowgill-stripe and @crioux-stripe are out for the next few weeks. to ensure we make progress on this, i am stepping in to help out here. could you please add me to the collaborator list? thanks!
> 
> @kmg-stripe could you ask cody to send me an email to verify your github id? Sorry just to be cautious as the write permission is pretty powerful.

@Andyz26 understood.  i just sent cody a message.  he is OOO for a few weeks, so hoping he happens to see it while he is out.

crioux-stripe commented:
Closing, as @kmg-stripe merged in a version of @mcowgill-stripe 's work in https://github.com/Netflix/mantis/pull/682
